src/framebuffer/common.rs

```diff
@@ -97,14 +97,32 @@ impl color {
         //    green   : offset = 5,   length =6,      msb_right = 0
         //    blue    : offset = 0,   length =5,      msb_right = 0
         //
-        let r5 = (u16::from(r8) >> 3) as u8;
-        let g6 = (u16::from(g8) >> 2) as u8;
-        let b5 = (u16::from(b8) >> 3) as u8;
-
-        [
-            (((g6 & 0b00_0111) << 5) | b5),
-            ((r5 << 3) | ((g6 & 0b11_1000) >> 3)),
-        ]
+        let r5 = (r8 as u16 + 1) * 0b11111 / 255;
+        let g6 = (g8 as u16 + 1) * 0b111111 / 255;
+        let b5 = (b8 as u16 + 1) * 0b11111 / 255;
+
+        let rgb565 = b5 << 11 | g6 << 5 | r5;
+
+        rgb565.to_le_bytes()
+    }
+}
+
+#[test]
+fn rgb565_conversions() {
+    // Ensure that min and max values are transformed faithfully
+    assert_eq!(color::RGB(0, 0, 0).to_rgb565(), [0, 0]);
+    assert_eq!(color::RGB(255, 255, 255).to_rgb565(), [255, 255]);
+    assert_eq!(color::from_native([0, 0]).to_rgb8(), [0, 0, 0]);
+    assert_eq!(color::from_native([255, 255]).to_rgb8(), [255, 255, 255]);
+    assert_eq!(color::GRAY(0).to_rgb565(), [255, 255]);
+    assert_eq!(color::GRAY(255).to_rgb565(), [0, 0]);
```

Comment:
We should eventually add some test here that ensures we don't get the color channels swapped... catch any future bugs like the ones you fixed in the last couple commits. Up to you whether you want to do that before merging or not - this is obviously a lot better than what we had before.

Reply:
I can hardcode RGB values for a couple of the built-in colors like RED, if that works? :)

Reply:
Works for me!

Reply:
I'll also throw in an `assert_eq!(color::RGB(255, 0, 0).to_rgb8(), [255, 0, 0])` for good measure

Reply:
Turns out I still need to figure out some endianness stuff after all. As long as `color::RED` is accurate and those native components actually correspond to red... I'll figure it out! lol

Reply:
Tests have been added, algos corrected for the third (fourth?) time